services/st_util/StUartBootloaderCommandHandler.cpp

```diff
@@ -0,0 +1,552 @@
+    {
+        commandActions.emplace_back(infra::InPlaceType<TransmitWithTwosComplementChecksum>(), *this, erase);
+        commandActions.emplace_back(infra::InPlaceType<ReceiveAckAction>(), *this);
+        commandActions.emplace_back(infra::InPlaceType<TransmitWithTwosComplementChecksum>(), *this, subcommand);
+        commandActions.emplace_back(infra::InPlaceType<ReceiveAckAction>(), *this);
+
+        timeout.Start(commandTimeout, [this]()
+            {
+                OnError("Timeout erasing memory");
+            });
+
+        ExecuteCommand(onDone);
+    }
+
+    void StUartBootloaderCommandHandler::Erase(uint8_t nPages, infra::ConstByteRange pages, const infra::Function<void()>& onDone)
+    {
+        if (pages.size() > 256)
+            std::abort();
+
+        internalBuffer[0] = static_cast<uint8_t>(pages.size() + 1);
+        std::copy(pages.begin(), pages.begin() + pages.size(), internalBuffer.begin() + 1);
+        internalRange = infra::MakeRange(internalBuffer.begin(), internalBuffer.begin() + internalBuffer[0]);
+
+        commandActions.emplace_back(infra::InPlaceType<TransmitWithTwosComplementChecksum>(), *this, erase);
+        commandActions.emplace_back(infra::InPlaceType<ReceiveAckAction>(), *this);
+        commandActions.emplace_back(infra::InPlaceType<TransmitWithChecksumAction>(), *this, internalRange);
+        commandActions.emplace_back(infra::InPlaceType<ReceiveAckAction>(), *this);
+
+        timeout.Start(commandTimeout, [this]()
+            {
+                OnError("Timeout erasing memory");
+            });
+
+        ExecuteCommand(onDone);
+    }
+
+    void StUartBootloaderCommandHandler::ExtendedErase(uint16_t subcommand, const infra::Function<void()>& onDone)
+    {
+        this->subcommand = subcommand;
+
+        commandActions.emplace_back(infra::InPlaceType<TransmitWithTwosComplementChecksum>(), *this, extendedErase);
+        commandActions.emplace_back(infra::InPlaceType<ReceiveAckAction>(), *this);
+        commandActions.emplace_back(infra::InPlaceType<TransmitWithChecksumAction>(), *this, infra::MakeConstByteRange(this->subcommand));
+        commandActions.emplace_back(infra::InPlaceType<ReceiveAckAction>(), *this);
+
+        timeout.Start(commandTimeout, [this]()
+            {
+                OnError("Timeout extended erasing memory");
+            });
+
+        ExecuteCommand(onDone);
+    }
+
+    void StUartBootloaderCommandHandler::ExtendedErase(uint8_t nPages, infra::ConstByteRange pages, const infra::Function<void()>& onDone)
+    {
+        if (pages.size() > 256)
+            std::abort();
+
+        internalBuffer[0] = static_cast<uint8_t>(nPages >> 8);
+        internalBuffer[1] = static_cast<uint8_t>(nPages & 0xff) + 1;
+        std::copy(pages.begin(), pages.begin() + pages.size(), internalBuffer.begin() + 2);
+        internalRange = infra::MakeRange(internalBuffer.begin(), internalBuffer.begin() + pages.size() + 2);
```

Comment:
nPages is a uint8_t, so nPages >> 8 is 0.

If internalBuffer is a BoundedVector instead of an std::array, then you can do this:
If you use a BoundedVectorOutStream to fill internalBuffer, then you can refactor this like this:
```suggestion
        internalBuffer.clear();
        infra::BoundedVectorOutputStream stream(internalBuffer);
        internalBuffer << static_cast<uint8_t>(nPages >> 8) << static_cast<uint8_t>((nPages & 0xff) + 1) << pages;
```